src/main/java/de/tum/in/www1/artemis/service/competency/CourseCompetencyService.java

```diff
@@ -63,6 +114,291 @@ public CourseCompetencyService(CompetencyProgressRepository competencyProgressRe
+
+        Set<Exercise> exercisesUserIsAllowedToSee = exerciseService.filterOutExercisesThatUserShouldNotSee(competency.getExercises(), currentUser);
+        Set<Exercise> exercisesWithAllInformationNeeded = exerciseService
+                .loadExercisesWithInformationForDashboard(exercisesUserIsAllowedToSee.stream().map(Exercise::getId).collect(Collectors.toSet()), currentUser);
+        competency.setExercises(exercisesWithAllInformationNeeded);
+    }
+
+    /**
+     * Imports the given course competencies and relations into a course
+     *
+     * @param course             the course to import into
+     * @param courseCompetencies the course competencies to import
+     * @return The set of imported course competencies, each also containing the relations it is the tail competency for.
+     */
+    public Set<CompetencyWithTailRelationDTO> importCourseCompetenciesAndRelations(Course course, Collection<CourseCompetency> courseCompetencies) {
+        var idToImportedCompetency = new HashMap<Long, CompetencyWithTailRelationDTO>();
+
+        for (var courseCompetency : courseCompetencies) {
+            CourseCompetency importedCompetency = switch (courseCompetency) {
+                case Competency competency -> new Competency(competency);
+                case Prerequisite prerequisite -> new Prerequisite(prerequisite);
+                default -> throw new IllegalStateException("Unexpected value: " + courseCompetency);
+            };
+            importedCompetency.setCourse(course);
+
+            importedCompetency = courseCompetencyRepository.save(importedCompetency);
+            idToImportedCompetency.put(courseCompetency.getId(), new CompetencyWithTailRelationDTO(importedCompetency, new ArrayList<>()));
+        }
+
+        return importCourseCompetenciesAndRelations(course, idToImportedCompetency);
+    }
+
+    /**
+     * Imports the given competencies and relations into a course
+     *
+     * @param course                 the course to import into
+     * @param idToImportedCompetency map of original competency id to imported competency
+     * @return The set of imported competencies, each also containing the relations it is the tail competency for.
+     */
+    public Set<CompetencyWithTailRelationDTO> importCourseCompetenciesAndRelations(Course course, Map<Long, CompetencyWithTailRelationDTO> idToImportedCompetency) {
+        if (course.getLearningPathsEnabled()) {
+            var importedCompetencies = idToImportedCompetency.values().stream().map(CompetencyWithTailRelationDTO::competency).toList();
+            learningPathService.linkCompetenciesToLearningPathsOfCourse(importedCompetencies, course.getId());
+        }
+
+        var originalCompetencyIds = idToImportedCompetency.keySet();
+        var relations = competencyRelationRepository.findAllByHeadCompetencyIdInAndTailCompetencyIdIn(originalCompetencyIds, originalCompetencyIds);
+
+        if (!relations.isEmpty()) {
```

Comment:
Is this check needed? If `relations` were empty, the forEach loop should not be executed once.